libcontainer/rootfs_linux.go

```diff
@@ -563,48 +563,66 @@ func setupPtmx(config *configs.Config, console *linuxConsole) error {
 	return nil
 }
 
-func pivotRoot(rootfs, pivotBaseDir string) (err error) {
-	if pivotBaseDir == "" {
-		pivotBaseDir = "/"
-	}
-	tmpDir := filepath.Join(rootfs, pivotBaseDir)
-	if err := os.MkdirAll(tmpDir, 0755); err != nil {
-		return fmt.Errorf("can't create tmp dir %s, error %v", tmpDir, err)
+// pivotRoot will call pivot_root such that rootfs becomes the new root
+// filesystem, and everything else is cleaned up.
+func pivotRoot(rootfs string) error {
+	// While the documentation may claim otherwise, pivot_root(".", ".") is
+	// actually valid. What this results in is / being the new root but
+	// /proc/self/cwd being the old root. Since we can play around with the cwd
+	// with pivot_root this allows us to pivot without creating directories in
+	// the rootfs. Shout-outs to the LXC developers for giving us this idea.
+
+	oldroot, err := syscall.Open("/", syscall.O_DIRECTORY|syscall.O_RDONLY, 0)
+	if err != nil {
+		return err
 	}
-	pivotDir, err := ioutil.TempDir(tmpDir, ".pivot_root")
+	defer syscall.Close(oldroot)
+
+	newroot, err := syscall.Open(rootfs, syscall.O_DIRECTORY|syscall.O_RDONLY, 0)
 	if err != nil {
-		return fmt.Errorf("can't create pivot_root dir %s, error %v", pivotDir, err)
+		return err
 	}
-	defer func() {
-		errVal := os.Remove(pivotDir)
-		if err == nil {
-			err = errVal
-		}
-	}()
-	if err := syscall.PivotRoot(rootfs, pivotDir); err != nil {
+	defer syscall.Close(newroot)
+
+	// Change to the new root so that the pivot_root actually acts on it.
+	if err := syscall.Fchdir(newroot); err != nil {
+		return err
+	}
+
+	if err := syscall.PivotRoot(".", "."); err != nil {
 		// Make the parent mount private
-		if err := rootfsParentMountPrivate(rootfs); err != nil {
+		if err := rootfsParentMountPrivate("."); err != nil {
 			return err
 		}
+
 		// Try again
-		if err := syscall.PivotRoot(rootfs, pivotDir); err != nil {
+		if err := syscall.PivotRoot(".", "."); err != nil {
 			return fmt.Errorf("pivot_root %s", err)
 		}
 	}
-	if err := syscall.Chdir("/"); err != nil {
-		return fmt.Errorf("chdir / %s", err)
+
+	// Currently our "." is oldroot (according to the current kernel code).
```

Comment:
I don't think you're right (I thought about it some more). `pivot_root(a, b)` will make `a` the new root and mount the old root at `b`. What we're doing here is unmounting the old root, because the old root is also on `b`. Since `b` is `.` it matters whether or not `pivot_root` has changed your `cwd`. It turns out that this does change your `cwd` in the current Linux implementation, but I don't want to depend on that (the mount logic is very dodgy in the kernel).